src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs

```diff
@@ -162,11 +197,139 @@ public Task BeforeStartAsync(DistributedApplicationModel appModel, CancellationT
                         }
                     }));
 
+            resource.Annotations.Add(
+                new ManifestPublishingCallbackAnnotation(
+                    writer =>
+                    {
+                        writer.WriteString("type", "dapr.v0");
+                        writer.WriteStartObject("dapr");
+                        writer.WriteString("application", project.Name);
+                        if (componentReferenceAnnotations.Any())
+                        {
+                            writer.WriteStartArray("components");
+                            foreach (var componentReferenceAnnotation in componentReferenceAnnotations)
+                            {
+                                writer.WriteStringValue(componentReferenceAnnotation.Component.Name);
+                            }
+                            writer.WriteEndArray();
+                        }
+                        writer.WriteEndObject();
+                    }));
+
             appModel.Resources.Add(resource);
         }
+    }
+
+    public async Task AfterStopAsync(CancellationToken cancellationToken = default)
+    {
+        await StopOnDemandDaprComponentsAsync().ConfigureAwait(false);
+    }
+
+    private async Task<IImmutableDictionary<string, string>> StartOnDemandDaprComponentsAsync(DistributedApplicationModel appModel, CancellationToken cancellationToken)
```

Comment:
Why an Immutable anything?

Reply:
Mostly habit. Switched to more common `IReadOnlyDictionary<string, string>`.